### PR TITLE
Fix a few low-hanging errors

### PR DESCRIPTION
**Pull request by bengotow**
This diff also fixes the following minor things:
Sentry #24197: 	Lots of events have end < start, just flip em and carry on
Sentry #5907:        	IMAP servers can return `nil` for internaldate. Handle alongside other bad input scenarios.
Sentry #7654:        Silently return (rather than throwing and retrying) if we try to syncback folder CRUD actions and the folder is gone.

`Category.display_name` typically contains the imap prefix for folders (INBOX.Foo for Fastmail.) API requests change this value to Foo when doing creates / renames, which is incorrect for that column. Previously, we were fixing and re-saving the folder from the sync code, which seemed like the wrong place to do it. Instead, just never put un-prefixed `Foo` into `Category.display_name` in the first place.

Test Plan: Changed a few tests, existing test suite still passes

### CODE REVIEW

File: inbox/api/ns_api.py [hunk 1]
```diff
@@ -532,6 +533,14 @@ def folders_labels_create_api():
     valid_display_name(g.namespace.id, category_type, display_name,
                        g.db_session)
 
+    if g.namespace.account.provider not in ['gmail', 'eas']:
```
Comment (khamidou) [inline inbox/api/ns_api.py]: Why do this?

File: inbox/api/ns_api.py [hunk 2]
```diff
@@ -581,6 +590,14 @@ def folder_label_update_api(public_id):
...
```
Comment (khamidou) [inline inbox/api/ns_api.py]: Same.

File: inbox/api/ns_api.py [hunk 3]
```diff
@@ -532,6 +533,14 @@ def folders_labels_create_api():
...
+        # Translate the name of the folder to an actual IMAP name
+        # (e.g: "Accounting/Taxes" becomes "Accounting.Taxes")
+        display_name = imap_folder_path(
```
Comment (khamidou) [inline inbox/api/ns_api.py]: Makes sense, but can you check what happens if you pass an IMAP path to `imap_folder_path`?

File: inbox/contacts/icloud.py [hunk 4]
```diff
@@ -13,7 +13,7 @@
 from inbox.models.backends.generic import GenericAccount
 
 
-ICLOUD_CONTACTS_URL = 'https://contacts.icloud.com/'
+ICLOUD_CONTACTS_URL = 'https://contacts.icloud.com'
```
Comment (khamidou) [inline inbox/contacts/icloud.py]: Isn't the problem that we're not joining URLs correctly below (e.g: line 79, 84 & al)?

File: inbox/mailsync/backends/imap/generic.py [hunk 5]
```diff
@@ -471,6 +471,13 @@ def create_message(self, db_session, acct, folder, msg):
                       existing_imapuid=existing_imapuid.id)
             return None
 
+        # Check if the message is valid. In some scenarios, we get messages with
+        # the date set to 1970 or a `None` body string.
+        # https://sentry.nylas.com/sentry/sync-prod/group/3387/
+        if msg.body is None or msg.internaldate == datetime(1970, 1, 1, 0, 0):
```
Comment (khamidou) [inline inbox/mailsync/backends/imap/generic.py]: A message with a date set to epoch could have contents though.

File: inbox/util/concurrency.py [hunk 6]
```diff
@@ -65,8 +71,17 @@ def wrapped(*args, **kwargs):
 def retry_with_logging(func, logger=None, retry_classes=None,
                        fail_classes=None, account_id=None, provider=None,
                        backoff_delay=BACKOFF_DELAY):
-    def callback():
-        log_uncaught_errors(logger, account_id=account_id, provider=provider)
+    network_errs = [0]
+
+    def callback(e):
+        if isinstance(e, TRANSIENT_NETWORK_ERRS):
+            network_errs[0] += 1
```
Comment (khamidou) [inline inbox/util/concurrency.py]: Are you resetting this anywhere? What happens for example if an IMAP account meets 20 transient SSL errors in the course of a regular run.

File: inbox/models/when.py [hunk 7]
```diff
@@ -97,7 +97,7 @@ def parse(cls, raw):
         # the singular form if they are equal.
         start, end = cls.parse_keys(raw)
         if start > end:
-            raise ValueError("'{}' must be < '{}'.".format(*cls.json_keys))
+            start, end = (end, start)
```
Comment (khamidou) [inline inbox/models/when.py]: Why fix this? We shouldn't try to second-guess API users. Also, do you have a link to the sentry trace --- I couldn't find it anywhere.

**Comment by khamidou** [comment]
Thanks for looking into this! I've left some comments inline. Also, next time can you break this into multiple PRs? It's very hard to review as-is.

File: inbox/contacts/icloud.py (hunk 4 above)
Comment (bengotow) [inline inbox/contacts/icloud.py]: Yeah - the problem is that we're just concatenating strings to build URLs, and at some point I assume the server started returning a URL fragment with the leading `/` on it. We could also fix it that way, but I've heard we may want to just delete all this code?

File: inbox/models/when.py (hunk 7 above)
Comment (bengotow) [inline inbox/models/when.py]: Oh sorry it's here https://sentry.nylas.com/sentry/sync-prod/group/24197/.

We could also skip syncing the event and ignore it - totally up to you. I'm not sure that the `start < end` rule exists on the remote system though, I think we just made it up for ourselves when we built this `when` wrapper around time formats.

**Comment by bengotow** [comment]
Yep can definitely break into multiple PRs next time—this one grew a bit. Just didn't want to make a bunch of one-line PRs. Will probably land them without squashing also so the individual sentry refs are preserved in the commit log.

File: inbox/api/ns_api.py (hunk 3 above)
Comment (bengotow) [inline inbox/api/ns_api.py]: Hmm so whatever the behavior is, this diff doesn't change it, just moves the call earlier into the initial save instead of a subsequent save. 

It looks like `imap_folder_path` checks that the path does not already start with the prefix though, so we should be good. (https://github.com/nylas/sync-engine/blob/master/inbox/util/misc.py#L220)

File: inbox/util/concurrency.py (hunk 6 above)
Comment (bengotow) [inline inbox/util/concurrency.py]: Good call — right now if you have `transient error, real error, transient error` it resets, but not if you have `transient error.... success for a long time .... transient error`. willfix.

File: inbox/mailsync/backends/imap/generic.py (hunk 5 above)
Comment (bengotow) [inline inbox/mailsync/backends/imap/generic.py]: Hmm true, but I'm not sure we should be trying to sync either of them? I grouped both of these conditions together because it seemed like either one would be a sign something is not right and this message is borked. We could just look at `msg.body is None` though, because that's the only one that causes an exception downstream.

File: inbox/util/concurrency.py (hunk 6 above)
Comment (bengotow) [inline inbox/util/concurrency.py]: [I'm going to break this change out into a separate PR]

File: inbox/mailsync/backends/imap/generic.py (hunk 5 above)
Comment (bengotow) [inline inbox/mailsync/backends/imap/generic.py]: Will just check for msg.body is None

**Comment by spang** [comment]
Cool, LGTM. Thanks for trawling through Sentry and fixing this stuff!
